wc/TryingSingletons.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE
+   DataKinds,
+   FlexibleContexts,
+   FlexibleInstances,
+   GADTs,
+   MultiParamTypeClasses,
+   PartialTypeSignatures,
+   RankNTypes,
+   ScopedTypeVariables,
+   StandaloneDeriving,
+   StandaloneKindSignatures,
+   TemplateHaskell,
+   TypeApplications,
+   TypeFamilies,
+   TypeOperators,
+   UndecidableInstances #-}
+
+module TryingSingletons where
+
+import Data.Singletons
+import Data.Singletons.TH
+import Data.List.Singletons
+import Text.Show.Singletons
+import Data.List (foldl')
+import Data.Int
+
+import Wc
+
+$(genSingletons [''CountMode])
+$(showSingInstances [''CountMode])
+
+$(genDefunSymbols [''CountBy])
+
+foo :: forall ms r. (CountModeC (CountByModes ms), PairList (Result (CountByModes ms)) Int64) =>
+    Sing ms -> String -> [Int64]
+foo modes = toList . getResult @(CountByModes ms) .
+    foldl' (\count char -> count <> fromChar @(CountByModes ms) char) mempty
+
+foo' :: [CountMode] -> String -> [Int64]
+foo' modes = case toSing modes of
+    SomeSing sms -> undefined
```

Comment:
the problem here is I should be able to just do
```haskell
foo' modes = case toSing modes of
    SomeSing sms -> foo sms
```
or even just
```haskell
foo' modes = withSomeSing modes foo
```
But it seems like the type checker isn't _quite_ convinced that the constraints on `foo` are valid for all possible `[CountMode]`. I think I can resolve that by rearranging some classes or something though. Maybe rather than using the `CountModeC` instances for pairs directly, and doing the `ListToPairs` thing, I should make a separate class that takes a `[*]` and gives back a Constraint that proves it can count up and give back an `[Int64]`...